### DISCUSSION TITLE
Implement spawn mechanism without reflection?

### DISCUSSION THREAD

**Discussion by chengenzhao** [AlmasB/FXGL#1126] [Ideas]
Hi:

Currently if need to use graalvm to aot our program written in FXGL
we have to write those class using reflection in XML 
is it possible to provide a way writing codes without xml configuration and annotation?
Which means FXGL needs to provide an implementation of spawn etc. without using reflection
so could it be possible to provide an alternative option for spawn methods?

Thanks

**Comment by AlmasB** [comment]
Good suggestion. Feel free to turn this into an issue and I'll schedule for FXGL 17 as I believe the implementation is quite trivial -- allow the dev to add spawn methods manually to some map:

`Map<String, Function<SpawnData, Entity> >`

None of the existing API will change.

**Reply by chengenzhao** [reply]
https://github.com/AlmasB/FXGL/issues/1127

submitted 

Cheers

**Reply by AlmasB** [reply]
Thanks, just thinking about this... The reflection mostly happens with EngineService and Component objects. Here's an example [pom.xml](https://github.com/AlmasB/FXGLGames/blob/3c4686ae624181bbad1e69f656d175253b2a9868/Mario/pom.xml#L132). So actually implementing spawn via API and not reflection won't solve much -- we possibly won't need to include the Factory class, but all other xml entries will still be required.

**Reply by chengenzhao** [reply]
Good morning
thanks for the reply
Seems like it requires more works
so is it possible to reimplement EngineService and Component objects without reflection as well?
This may require more work but we could implement this step by step, starting by reimplement spawn method, and few versions later, remove all configs
